### PR TITLE
Keep node local dns components during the roll out of new nodes.

### PR DESCRIPTION
**Pull request by ScheererJ**
**How to categorize this PR?**

/area networking
/kind enhancement

**What this PR does / why we need it**:
Keep node local dns components during the roll out of new nodes.

DNS is crucial to kubernetes. Therefore, it is important that the node local
dns components stay in place on the (old) nodes until the node is being rolled.
With this change, the node local dns components stay in the cluster until no
node has the corresponding label requiring node local dns. In the following
reconcile the resources will be removed.
(The pods are only scheduled on the proper nodes using a node selector already.)

**Which issue(s) this PR fixes**:
None.

**Special notes for your reviewer**:
The majority of the work was already done in https://github.com/gardener/gardener/pull/6057.

**Release note**:

```other operator
Node local dns components will stay in the cluster until after the node roll out of a node local dns switch and be cleaned up in the next reconcile.
```

### CODE REVIEW

**Comment by ScheererJ** [comment]
Adding tests later today.

File: pkg/operation/botanist/nodelocaldns.go [hunk 1]
```diff
@@ -16,11 +16,14 @@ package botanist
...
 import (
 	"context"
+	"strconv"
 
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
 	v1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
 	"github.com/gardener/gardener/pkg/operation/botanist/component/nodelocaldns"
 	"github.com/gardener/gardener/pkg/utils/images"
 	"github.com/gardener/gardener/pkg/utils/imagevector"
+	corev1 "k8s.io/api/core/v1"
```
Comment (rfranzke) [inline pkg/operation/botanist/nodelocaldns.go:27]: Nit:

```suggestion

	corev1 "k8s.io/api/core/v1"
```

File: pkg/operation/botanist/nodelocaldns.go [hunk 2]
```diff
@@ -59,5 +62,28 @@ func (b *Botanist) ReconcileNodeLocalDNS(ctx context.Context) error {
 	if b.Shoot.NodeLocalDNSEnabled {
 		return b.Shoot.Components.SystemComponents.NodeLocalDNS.Deploy(ctx)
 	}
+	if stillDesired, err := b.isNodeLocalDNSStillDesired(ctx); err != nil {
+		return err
+	} else if stillDesired {
+		// Leave NodeLocalDNS components in the cluster until all nodes have been rolled
+		return nil
+	}
```
Comment (rfranzke) [inline pkg/operation/botanist/nodelocaldns.go:74]: Nit: formatting/readability:
```suggestion

	if stillDesired, err := b.isNodeLocalDNSStillDesired(ctx); err != nil {
		return err
	} else if stillDesired {
		// Leave NodeLocalDNS components in the cluster until all nodes have been rolled
		return nil
	}

```

File: pkg/operation/botanist/nodelocaldns.go [hunk 3]
```diff
@@ -59,5 +62,28 @@ func (b *Botanist) ReconcileNodeLocalDNS(ctx context.Context) error {
...
+
+	for _, node := range nodeList.Items {
+		nodeLocalDNSLabelValue := node.Labels[v1beta1constants.LabelNodeLocalDNS]
+		if strconv.FormatBool(true) == nodeLocalDNSLabelValue {
+			return true, nil
+		}
+	}
+
+	return false, nil
```
Comment (rfranzke) [inline pkg/operation/botanist/nodelocaldns.go]: You could use a metadata-only list with a label selector for `v1beta1constants.LabelNodeLocalDNS=true` and `client.Limit(1)` instead to reduce the network I/O / stress on KAPI.

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/operation/botanist/nodelocaldns.go (hunk 1 above)
Comment (ScheererJ) [reply pkg/operation/botanist/nodelocaldns.go:27]: Addressed with latests commit.

File: pkg/operation/botanist/nodelocaldns.go (hunk 2 above)
Comment (ScheererJ) [reply pkg/operation/botanist/nodelocaldns.go:74]: Addressed with latests commit.

File: pkg/operation/botanist/nodelocaldns.go (hunk 3 above)
Comment (ScheererJ) [reply pkg/operation/botanist/nodelocaldns.go]: Addressed with latests commit.

File: pkg/operation/botanist/nodelocaldns.go [hunk 4]
```diff
@@ -59,5 +65,33 @@ func (b *Botanist) ReconcileNodeLocalDNS(ctx context.Context) error {
...
```
Comment (rfranzke) [inline pkg/operation/botanist/nodelocaldns.go]:
```suggestion

	return len(nodeList.Items) > 0, nil
```

Comment (ScheererJ) [reply pkg/operation/botanist/nodelocaldns.go]: Good point :-) .
Adapted accordingly.

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**